// minecraft-motd-util/dist/color.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ColorUtil = /** @class */ (function () {
    function ColorUtil(list) {
        this.list = [];
        for (var key in list) {
            this.list.push({
                name: key,
                hex: list[key],
                sum: sum(list[key])
            });
        }
    }
    ColorUtil.prototype.closest = function (input) {
        var colorSum = sum(input);
        var closest = null;
        var lastDifference = Infinity;
        for (var _i = 0, _a = this.list; _i < _a.length; _i++) {
            var color = _a[_i];
            var diff = Math.abs(colorSum - color.sum);
            if (closest === null || diff < lastDifference) {
                closest = color;
                lastDifference = diff;
            }
        }
        return closest;
    };
    return ColorUtil;
}());
var sum = function (input) {
    var sum = 0;
    input = input.replace('#', '');
    var r = input.substring(0, 2);
    var g = input.substring(2, 4);
    var b = input.substring(4, 6);
    sum = Math.sqrt(Math.pow(parseInt(r, 16), 2) + Math.pow(parseInt(g, 16), 2) + Math.pow(parseInt(b, 16), 2));
    return sum;
};
exports.default = ColorUtil;
