// minecraft-server-util/dist/sendVote.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.sendVote = void 0;
const assert_1 = __importDefault(require("assert"));
const crypto_1 = __importDefault(require("crypto"));
const util_1 = require("util");
const TCPClient_1 = __importDefault(require("./structure/TCPClient"));
const encoder = new util_1.TextEncoder();
function sendVote(host, port = 8192, options) {
    (0, assert_1.default)(typeof host === 'string', `Expected 'host' to be a 'string', got '${typeof host}'`);
    (0, assert_1.default)(host.length > 1, `Expected 'host' to have a length greater than 0, got ${host.length}`);
    (0, assert_1.default)(typeof port === 'number', `Expected 'port' to be a 'number', got '${typeof port}'`);
    (0, assert_1.default)(Number.isInteger(port), `Expected 'port' to be an integer, got '${port}'`);
    (0, assert_1.default)(port >= 0, `Expected 'port' to be greater than or equal to 0, got '${port}'`);
    (0, assert_1.default)(port <= 65535, `Expected 'port' to be less than or equal to 65535, got '${port}'`);
    (0, assert_1.default)(typeof options === 'object', `Expected 'options' to be an 'object', got '${typeof options}'`);
    (0, assert_1.default)(typeof options.username === 'string', `Expected 'options.username' to be an 'string', got '${typeof options.username}'`);
    (0, assert_1.default)(options.username.length > 1, `Expected 'options.username' to have a length greater than 0, got ${options.username.length}`);
    (0, assert_1.default)(typeof options.token === 'string', `Expected 'options.token' to be an 'string', got '${typeof options.token}'`);
    (0, assert_1.default)(options.token.length > 1, `Expected 'options.token' to have a length greater than 0, got ${options.token.length}`);
    return new Promise((resolve, reject) => __awaiter(this, void 0, void 0, function* () {
        var _a, _b, _c, _d;
        let socket = undefined;
        const timeout = setTimeout(() => {
            socket === null || socket === void 0 ? void 0 : socket.close();
            reject(new Error('Timed out while retrieving server status'));
        }, (_a = options === null || options === void 0 ? void 0 : options.timeout) !== null && _a !== void 0 ? _a : 1000 * 5);
        try {
            socket = new TCPClient_1.default();
            yield socket.connect({ host, port, timeout: (_b = options === null || options === void 0 ? void 0 : options.timeout) !== null && _b !== void 0 ? _b : 1000 * 5 });
            let challengeToken;
            // Handshake packet
            // https://github.com/NuVotifier/NuVotifier/wiki/Technical-QA#handshake
            {
                const version = yield socket.readStringUntil(0x0A);
                const split = version.split(' ');
                if (split[1] !== '2')
                    throw new Error('Unsupported Votifier version: ' + split[1]);
                challengeToken = split[2];
            }
            // Send vote packet
            // https://github.com/NuVotifier/NuVotifier/wiki/Technical-QA#protocol-v2
            {
                const payload = {
                    serviceName: (_c = options.serviceName) !== null && _c !== void 0 ? _c : 'minecraft-server-util (https://github.com/PassTheMayo/minecraft-server-util)',
                    username: options.username,
                    address: host + ':' + port,
                    timestamp: (_d = options.timestamp) !== null && _d !== void 0 ? _d : Date.now(),
                    challenge: challengeToken
                };
                if (options.uuid) {
                    payload.uuid = options.uuid;
                }
                const payloadSerialized = JSON.stringify(payload);
                const message = {
                    payload: payloadSerialized,
                    signature: crypto_1.default.createHmac('sha256', options.token).update(payloadSerialized).digest('base64')
                };
                const messageSerialized = JSON.stringify(message);
                const messageBytes = encoder.encode(messageSerialized);
                socket.writeInt16BE(0x733A);
                socket.writeInt16BE(messageBytes.byteLength);
                socket.writeBytes(messageBytes);
                yield socket.flush(false);
            }
            // Response packet
            // https://github.com/NuVotifier/NuVotifier/wiki/Technical-QA#protocol-v2
            {
                const responseString = yield socket.readStringUntil(0x0A);
                const response = JSON.parse(responseString);
                socket.close();
                clearTimeout(timeout);
                switch (response.status) {
                    case 'ok': {
                        resolve();
                        break;
                    }
                    case 'error': {
                        reject(new Error(response.cause + ': ' + response.error));
                        break;
                    }
                    default: {
                        reject(new Error('Server sent an unknown response: ' + responseString));
                        break;
                    }
                }
            }
        }
        catch (e) {
            clearTimeout(timeout);
            socket === null || socket === void 0 ? void 0 : socket.close();
            reject(e);
        }
    }));
}
exports.sendVote = sendVote;
