// minecraft-server-util/dist/structure/RCON.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RCON = void 0;
const assert_1 = __importDefault(require("assert"));
const events_1 = require("events");
const util_1 = require("util");
const TCPClient_1 = __importDefault(require("./TCPClient"));
const encoder = new util_1.TextEncoder();
class RCON extends events_1.EventEmitter {
    constructor() {
        super();
        this.isLoggedIn = false;
        this.socket = null;
        this.requestID = 0;
    }
    get isConnected() {
        return this.socket && this.socket.isConnected || false;
    }
    connect(host, port = 25575, options = {}) {
        (0, assert_1.default)(typeof host === 'string', `Expected 'host' to be a 'string', got '${typeof host}'`);
        (0, assert_1.default)(host.length > 1, `Expected 'host' to have a length greater than 0, got ${host.length}`);
        (0, assert_1.default)(typeof port === 'number', `Expected 'port' to be a 'number', got '${typeof port}'`);
        (0, assert_1.default)(Number.isInteger(port), `Expected 'port' to be an integer, got '${port}'`);
        (0, assert_1.default)(port >= 0, `Expected 'port' to be greater than or equal to 0, got '${port}'`);
        (0, assert_1.default)(port <= 65535, `Expected 'port' to be less than or equal to 65535, got '${port}'`);
        (0, assert_1.default)(typeof options === 'object', `Expected 'options' to be an 'object', got '${typeof options}'`);
        return new Promise((resolve, reject) => {
            var _a;
            this.socket = new TCPClient_1.default();
            const timeout = setTimeout(() => {
                var _a;
                reject(new Error('Failed to connect to RCON server within timeout duration'));
                (_a = this.socket) === null || _a === void 0 ? void 0 : _a.close();
            }, (_a = options === null || options === void 0 ? void 0 : options.timeout) !== null && _a !== void 0 ? _a : 1000 * 5);
            this.socket.connect(Object.assign({ host, port }, options))
                .then(() => {
                clearTimeout(timeout);
                resolve();
            })
                .catch((error) => {
                clearTimeout(timeout);
                reject(error);
            });
        });
    }
    login(password, options = {}) {
        (0, assert_1.default)(typeof password === 'string', `Expected 'password' to be a 'string', got '${typeof password}'`);
        (0, assert_1.default)(password.length > 1, `Expected 'password' to have a length greater than 0, got ${password.length}`);
        (0, assert_1.default)(typeof options === 'object', `Expected 'options' to be an 'object', got '${typeof options}'`);
        return new Promise((resolve, reject) => __awaiter(this, void 0, void 0, function* () {
            var _a;
            if (this.socket === null || !this.socket.isConnected)
                return reject(new Error('login() attempted before RCON has connected'));
            const timeout = setTimeout(() => {
                var _a;
                reject(new Error('Failed to connect to RCON server within timeout duration'));
                (_a = this.socket) === null || _a === void 0 ? void 0 : _a.close();
            }, (_a = options === null || options === void 0 ? void 0 : options.timeout) !== null && _a !== void 0 ? _a : 1000 * 5);
            this.isLoggedIn = false;
            const passwordBytes = encoder.encode(password);
            // Login packet
            // https://wiki.vg/RCON#3:_Login
            {
                this.socket.writeInt32LE(10 + passwordBytes.byteLength);
                this.socket.writeInt32LE(this.requestID++);
                this.socket.writeInt32LE(3);
                this.socket.writeBytes(passwordBytes);
                this.socket.writeBytes(Uint8Array.from([0x00, 0x00]));
                yield this.socket.flush(false);
            }
            // Login response packet
            // https://wiki.vg/RCON#3:_Login
            {
                const packetLength = yield this.socket.readInt32LE();
                this.socket.ensureBufferedData(packetLength);
                const requestID = yield this.socket.readInt32LE();
                if (requestID === -1)
                    throw new Error('Invalid RCON password');
                const packetType = yield this.socket.readInt32LE();
                if (packetType !== 2)
                    throw new Error('Expected server to send packet type 2, received ' + packetType);
                yield this.socket.readBytes(2);
            }
            this.isLoggedIn = true;
            clearTimeout(timeout);
            resolve();
            process.nextTick(() => __awaiter(this, void 0, void 0, function* () {
                while (this.socket !== null && this.socket.isConnected && this.isLoggedIn) {
                    try {
                        yield this._readPacket();
                    }
                    catch (e) {
                        this.emit('error', e);
                    }
                }
            }));
        }));
    }
    run(command) {
        return __awaiter(this, void 0, void 0, function* () {
            (0, assert_1.default)(typeof command === 'string', `Expected 'command' to be a 'string', got '${typeof command}'`);
            (0, assert_1.default)(command.length > 1, `Expected 'command' to have a length greater than 0, got ${command.length}`);
            if (this.socket === null || !this.socket.isConnected)
                throw new Error('run() attempted before RCON has connected');
            if (!this.isLoggedIn)
                throw new Error('run() attempted before RCON has successfully logged in');
            const commandBytes = encoder.encode(command);
            const requestID = this.requestID++;
            this.socket.writeInt32LE(10 + commandBytes.byteLength);
            this.socket.writeInt32LE(requestID);
            this.socket.writeInt32LE(2);
            this.socket.writeBytes(commandBytes);
            this.socket.writeBytes(Uint8Array.from([0x00, 0x00]));
            yield this.socket.flush(false);
            return requestID;
        });
    }
    execute(command) {
        return __awaiter(this, void 0, void 0, function* () {
            (0, assert_1.default)(typeof command === 'string', `Expected 'command' to be a 'string', got '${typeof command}'`);
            (0, assert_1.default)(command.length > 1, `Expected 'command' to have a length greater than 0, got ${command.length}`);
            const requestID = yield this.run(command);
            return new Promise((resolve) => {
                const listenerFunc = (data) => {
                    if (data.requestID !== requestID)
                        return;
                    this.removeListener('message', listenerFunc);
                    resolve(data.message);
                };
                this.on('message', listenerFunc);
            });
        });
    }
    _readPacket() {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.socket === null || !this.socket.isConnected || !this.isLoggedIn)
                return Promise.reject(new Error('Attempted to read packet when socket was disconnected or RCON was not logged in'));
            const packetLength = yield this.socket.readInt32LE();
            yield this.socket.ensureBufferedData(packetLength);
            const requestID = yield this.socket.readInt32LE();
            const packetType = yield this.socket.readInt32LE();
            if (packetType === 0) {
                const message = yield this.socket.readStringNT();
                yield this.socket.readBytes(1);
                this.emit('message', { requestID, message });
            }
            else {
                yield this.socket.readBytes(packetLength - 8);
            }
        });
    }
    close() {
        var _a;
        (_a = this.socket) === null || _a === void 0 ? void 0 : _a.close();
    }
}
exports.RCON = RCON;
